### PR TITLE
Features/admin applicationdetail service

### PR DESCRIPTION
**Pull request by Divya-Kaavian**

### CODE REVIEW

File: srv/dte-energy-admin-service.cds [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 service DTEEnergyAdminPortal {
 
-   entity ApplicationDetail as projection on db.ApplicationDetail;
-
+  entity ApplicationDetail as projection on db.ApplicationDetail actions{
+    function GenerateUrl() returns String;
+    action UpdateAppDetail(AppId: String, NoOfTenants: Integer, AppStatus: String) returns String;
+  };
+  
+  entity ApplicationConsent as projection on db.ApplicationConsent;
```
Comment (Kokila-Kaavian) [inline srv/dte-energy-admin-service.cds:12]: Update ConsentStatus Action Missing.

File: srv/handler/generate-consent-url.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Method to generate the consent url with the encrypted appId
+ * @param {String} appId 
+ * @returns {String} consentEncrLink
+ */
+const generateConsentUrl = async(appId, ApplicationDetail) =>{
+    // Get the Consent Service URL
+    const url = process.env.CONSENT_APP_URL;
+    if(appId === "") throw {message: 'AppId is empty', status: 400};
```
Comment (Kokila-Kaavian) [inline srv/handler/generate-consent-url.js:11]:
```suggestion
    if(appId === "") throw {message: 'Application ID is required', status: 400};
```

File: .env [hunk 3]
```diff
@@ -1,3 +1,4 @@
 CRYPTOJS_SECRET_KEY = "@dteEneHub*123*ConsentApp"
 HTML5_PRODUCTION = "https://port4004-workspaces-ws-b66tw.us20.applicationstudio.cloud.sap/dteconsentappclient/webapp/index.html"
-ADMIN_PRODUCTION = "https://port4004-workspaces-ws-b66tw.us20.applicationstudio.cloud.sap/dte-energy-admin-portal/webapp/index.html"
\ No newline at end of file
+ADMIN_PRODUCTION = "https://port4004-workspaces-ws-b66tw.us20.applicationstudio.cloud.sap/dte-energy-admin-portal/webapp/index.html"
+CONSENT_APP_URL = "https://fiori-dev-dte.launchpad.cfapps.us20.hana.ondemand.com/5cac032f-5754-422c-bdf1-b5c9543524d4.dteConsentAppPortal.dteconsentappclient-0.0.1/index.html#tenant/consent/?encrAppId="
```
Comment (Kokila-Kaavian) [inline .env:4]:
```suggestion
TENANT_CONSENT_FORM_URL = "https://fiori-dev-dte.launchpad.cfapps.us20.hana.ondemand.com/5cac032f-5754-422c-bdf1-b5c9543524d4.dteConsentAppPortal.dteconsentappclient-0.0.1/index.html#tenant/consent/?encrAppId="
```

File: srv/handler/generate-consent-url.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        
+    // Get the application detail
+    const appDetail = await SELECT.from(ApplicationDetail).columns('AppId').where({'AppId': appId});
+    
+    // If the length is empty
+    if(appDetail.length === 0) throw {'message':'Application Details for this Id is not found'}
```
Comment (Kokila-Kaavian) [inline srv/handler/generate-consent-url.js:17]:
```suggestion
    if(appDetail.length === 0) throw {'message':'Application Details is not found. Invalid Application ID'}
```

File: srv/handler/generate-consent-url.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    // Encrypt the appid
+    const encrAppId = await valueEncrypt(appId);
+
+    // Update the consent url with the encrypted id.
+    const consentEncrLink = url+encrAppId;
```
Comment (Kokila-Kaavian) [inline srv/handler/generate-consent-url.js:23]:
```suggestion
    const consentFormLink = url+encrAppId;
```

**Review by Kokila-Kaavian — APPROVED** [review]

File: .env
Comment (Kokila-Kaavian) [inline .env:1]: Resolve Conflict

**Review by Kokila-Kaavian — CHANGES_REQUESTED** [review]

File: srv/dte-energy-admin-service.cds (hunk 1 above)
Comment (Divya-Kaavian) [reply srv/dte-energy-admin-service.cds:12]: The updation of fields can be achieved by the OData model configuratioin. So this is no more needed here.

File: srv/handler/admin-handler.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            for(const e of data){
+                // Check the ReferenceId with the AppId
+                const ConsentDetailCount = await cds.run(
+                    SELECT.from(ApplicationConsent)
+                    .where({AppRefId_AppId: e.AppId})
+                )
+                e.NoOfConsentReceived =  ConsentDetailCount.length;
+            }
+        } else {
+            // Set the NoOdConsentReceived as 0
```
Comment (Kokila-Kaavian) [inline srv/handler/admin-handler.js]: There is no need of else block

File: srv/handler/admin-handler.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                return {'message':e.message, 'code':'500'}
+            }
+        }
+        
+    }),
+
+    // Method to add the NoOfConsentReceived Field.
+    srv.after('READ', 'ApplicationDetail', async(data)=>{
+        // If data contains value
+        if(Array.isArray(data)){
```
Comment (Kokila-Kaavian) [inline srv/handler/admin-handler.js]: add try catch

File: srv/handler/generate-consent-url.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+const { valueEncrypt } = require("./encrypt-and-decrypt-id");
+
+/**
+ * Method to generate the consent url with the encrypted appId
+ * @param {String} appId 
+ * @returns {String} consentEncrLink
+ */
+const generateConsentUrl = async(appId, ApplicationDetail) =>{
+    // Get the Consent Service URL
+    const url = process.env.TENANT_CONSENT_FORM_URL;
```
Comment (Kokila-Kaavian) [inline srv/handler/generate-consent-url.js:11]: add try catch

File: srv/handler/admin-handler.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  }),
+
+    // Method to add the NoOfConsentReceived Field.
+    srv.after('READ', 'ApplicationDetail', async (data) => {
+      try {
+        // If data contains value
+        if (Array.isArray(data)) {
+          for (const e of data) {
+            // Check the ReferenceId with the AppId
+            const ConsentDetailCount = await cds.run(
```
Comment (Kokila-Kaavian) [inline srv/handler/admin-handler.js]: give proper variable name

**Review by Kokila-Kaavian — CHANGES_REQUESTED** [review]

File: srv/handler/generate-consent-url.js (hunk 8 above)
Comment (Divya-Kaavian) [reply srv/handler/generate-consent-url.js:11]: Done

File: srv/handler/admin-handler.js (hunk 9 above)
Comment (Divya-Kaavian) [reply srv/handler/admin-handler.js]: Done

**Review by Kokila-Kaavian — APPROVED** [review]

**Pull request merged**